modules/siddhi-core/src/main/java/org/wso2/siddhi/core/aggregation/IncrementalDataPurging.java

```diff
@@ -0,0 +1,291 @@
+            annotationTypes.put(annotation.getName().toLowerCase(), annotation);
+        }
+        Annotation purge = annotationTypes.get(SiddhiConstants.NAMESPACE_PURGE);
+        if (Objects.nonNull(purge)) {
+            if (Objects.nonNull(purge.getElement(SiddhiConstants.ANNOTATION_ELEMENT_ENABLE))) {
+                String enable = purge.getElement(SiddhiConstants.ANNOTATION_ELEMENT_ENABLE);
+                if (!("true".equalsIgnoreCase(enable) || "false".equalsIgnoreCase(enable))) {
+                    throw new SiddhiAppCreationException("Undefined value for enable: " + enable + "." +
+                            " Please use true or false");
+                } else {
+                    purgingEnabled = Boolean.parseBoolean(enable);
+                }
+            }
+            if (purgingEnabled) {
+                // If interval is defined, default value of 15 min will be replaced by user input value
+                if (Objects.nonNull(purge.getElement(SiddhiConstants.NAMESPACE_INTERVAL))) {
+                    String interval = purge.getElement(SiddhiConstants.NAMESPACE_INTERVAL);
+                    purgeExecutionInterval = timeToLong(interval);
+                }
+                List<Annotation> retentions = purge.getAnnotations(SiddhiConstants.NAMESPACE_RETENTION);
+                if (Objects.nonNull(retentions) && !retentions.isEmpty()) {
+                    Annotation retention = retentions.get(0);
+                    List<Element> elements = retention.getElements();
+                    for (Element element : elements) {
+                        TimePeriod.Duration duration = normalizeDuration(element.getKey());
+                        if (!aggregationTables.keySet().contains(duration)) {
+                            throw new SiddhiAppCreationException(duration + " granularity cannot be purged since " +
+                                    "aggregation has not performed in " + duration + " granularity");
+                        }
+                        if (element.getValue().equalsIgnoreCase("all")) {
+                            retentionPeriods.put(duration, RETAIN_ALL);
+                        } else {
+                            retentionPeriods.put(duration, timeToLong(element.getValue()));
+                        }
+                    }
+                }
+            }
+        }
+        compiledConditionsHolder = createCompileConditions(aggregationTables, tableMap);
+    }
+
+    private Long getPurgeExecutionInterval() {
+        return purgeExecutionInterval;
+    }
+
+    public boolean isPurgingEnabled() {
+        return purgingEnabled;
+    }
+
+    @Override
+    public void run() {
+        long currentTime = System.currentTimeMillis();
+        long purgeTime;
+        Object[] purgeTimeArray = new Object[1];
+
+        for (Map.Entry<TimePeriod.Duration, Table> entry : aggregationTables.entrySet()) {
+            if (!retentionPeriods.get(entry.getKey()).equals(RETAIN_ALL)) {
```

Comment:
Are we sure that `retentionPeriods.get(entry.getKey())` always returns a non-null value?

Reply:
Yes, we are adding all the possible durations when initializing the retentionPeriods map. So it won't be null.